xds/internal/balancer/cdsbalancer/cdsbalancer_security_test.go

```diff
@@ -18,718 +18,841 @@ package cdsbalancer
+		SkipValidation: true,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := mgmtServer.Update(ctx, resources); err != nil {
 		t.Fatal(err)
 	}
 
-	// Make a NewSubConn and verify that attributes are added.
-	if _, err := makeNewSubConn(ctx, edsB.parentCC, tcc, false); err != nil {
-		t.Fatal(err)
+	// Verify that a successful RPC can be made.
+	client := testgrpc.NewTestServiceClient(cc)
+	peer := &peer.Peer{}
+	if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.WaitForReady(true), grpc.Peer(peer)); err != nil {
+		t.Fatalf("EmptyCall() failed: %v", err)
 	}
+	verifySecurityInformationFromPeer(t, peer, e2e.SecurityLevelMTLS)
 
-	// Here we invoke the watch callback registered on the fake xdsClient with
-	// an update which contains bad security config. So, we expect the CDS
-	// balancer to forward this error to the EDS balancer and eventually the
-	// channel needs to be put in a bad state.
-	xdsC.InvokeWatchClusterCallback(cdsUpdateWithMissingSecurityCfg, nil)
+	// Start a test service backend that does not expect a secure connection.
+	insecureServer := stubserver.StartTestService(t, nil)
+	t.Cleanup(insecureServer.Stop)
 
-	// We manually check that an error is forwarded to the EDS balancer instead
-	// of using one of the helper methods on the testEDSBalancer, because all we
-	// care here is whether an error is sent to it or not. We don't care about
-	// the exact error.
-	gotErr, err := edsB.resolverErrCh.Receive(ctx)
-	if err != nil {
-		t.Fatal("timeout waiting for CDS balancer to forward error to EDS balancer upon receipt of bad security config")
+	// Update the resources in the management server to contain no security
+	// configuration. This should result in the use of fallback credentials,
+	// which is insecure in our case.
+	resources = e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Clusters:       []*v3clusterpb.Cluster{e2e.DefaultCluster(clusterName, serviceName, e2e.SecurityLevelNone)},
+		Endpoints:      []*v3endpointpb.ClusterLoadAssignment{e2e.DefaultEndpoint(serviceName, "localhost", []uint32{testutils.ParsePort(t, insecureServer.Address)})},
+		SkipValidation: true,
 	}
-	if gotErr == nil {
-		t.Fatal("CDS balancer did not forward error to EDS balancer upon receipt of bad security config")
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
 	}
 
-	// Since the error being pushed here is not a resource-not-found-error, the
-	// registered watch should not be cancelled.
-	sCtx, sCancel := context.WithTimeout(context.Background(), defaultTestShortTimeout)
-	defer sCancel()
-	if _, err := xdsC.WaitForCancelClusterWatch(sCtx); err != context.DeadlineExceeded {
-		t.Fatal("cluster watch cancelled for a non-resource-not-found-error")
+	// Wait for the connection to move to the new backend that expects
+	// connections without security.
```

Comment:
The new cluster resource pushed from here does not contain any security information. Once cds receives this, it is expected to not use the fallback credentials, which in our case is insecure creds.

It was hard to get the same old server (which was expecting a secure connection) to now expect an insecure connection. Hence I spun up a new one, and changed the EDS resource also to point to the new one.

The reason I have a while loop here is because it can take some time for the security configuration to be received and processed by the cds LB policy. And the corresponding EDS resource to be processed by the child policies down the tree.